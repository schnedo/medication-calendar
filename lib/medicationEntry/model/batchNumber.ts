export default interface BatchNumber {
  readonly number: number;
}

const prefix = "Ch.-B.";
export function format(batchNumber: BatchNumber): string {
  return `${prefix} ${batchNumber.number}`;
}
