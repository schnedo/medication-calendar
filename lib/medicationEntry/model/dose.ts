export default interface Dose {
  readonly amount: number;
}

const unit = "ml";

export function format(dose: Dose): string {
  return `${dose.amount} ${unit}`;
}
