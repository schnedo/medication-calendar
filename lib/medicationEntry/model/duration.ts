export default interface Duration {
  minutes: number;
  hours: number;
}

function formatPart(number: number | undefined) {
  if (number === undefined) {
    return "00";
  }
  const raw = number.toString();
  return raw.length === 1 ? `0${raw}` : raw;
}

export function format(duration: Duration): string {
  return `${formatPart(duration.hours)}:${formatPart(duration.minutes)}`;
}
