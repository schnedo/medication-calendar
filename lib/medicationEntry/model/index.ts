export { default as BatchNumber } from "./batchNumber";
export { default as Dose } from "./dose";
export { default as Duration } from "./Duration";
export { default as Medicament } from "./medicament";
export { default as Medication } from "./medication";
export { default as MedicationEntry } from "./medicationEntry";

export * as Repository from "./repository";
