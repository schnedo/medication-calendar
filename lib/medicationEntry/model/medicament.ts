import BatchNumber from "./batchNumber";

export default interface Medicament {
  readonly name: string;
  readonly batchNumber: BatchNumber;
}
