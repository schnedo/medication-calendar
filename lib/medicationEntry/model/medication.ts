import Dose from "./dose";
import Medicament from "./medicament";
import { v4 } from "uuid";

export default class Medication {
  readonly id: string = v4();

  constructor(readonly dose: Dose, readonly medicament: Medicament) {}
}
