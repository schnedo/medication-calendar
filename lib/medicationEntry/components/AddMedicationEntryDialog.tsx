import { MouseEventHandler, ReactElement, useState } from "react";
import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  makeStyles,
  TextField,
} from "@material-ui/core";
import { DatePicker } from "@material-ui/pickers";
import { ModalProps } from "@material-ui/core/Modal";
import { Duration, Medication, MedicationEntry } from "../model";
import DurationInput from "./DurationInput";
import MedicationsInput from "./MedicationsInput";

const useStyles = makeStyles((theme) => ({
  dialogContent: {
    display: "flex",
    flexDirection: "column",
    "&>*": {
      marginBottom: theme.spacing(2),
      "&:lastChild": {
        marginBottom: theme.spacing(0),
      },
    },
  },
}));

export interface AddMedicationEntryDialogProps {
  open: boolean;
  onClose?: ModalProps["onClose"];
  onSubmit?: (entry: MedicationEntry) => Promise<void>;
  onAbort?: () => void;
}

export default function AddMedicationEntryDialog({
  onClose,
  onSubmit,
  onAbort,
  open,
}: AddMedicationEntryDialogProps): ReactElement {
  const { dialogContent } = useStyles();

  const [date, setDate] = useState(new Date());
  const [duration, setDuration] = useState({
    minutes: 0,
    hours: 0,
  } as Duration);
  const [bodyMass, setBodyMass] = useState("60");
  const [comments, setComments] = useState("");
  const [medications, setMedications] = useState<Medication[]>([]);

  const handleSubmit: MouseEventHandler | undefined = onSubmit
    ? async () => {
        await onSubmit({
          id: "",
          date,
          duration,
          bodyMass: {
            amount: parseInt(bodyMass),
          },
          comments,
          medications,
        });
      }
    : undefined;

  return (
    <Dialog open={open} onClose={onClose}>
      <DialogTitle>Neuer Eintrag</DialogTitle>
      <DialogContent className={dialogContent}>
        <TextField
          id={"bodyMassInput"}
          label={"KÃ¶rpergewicht"}
          type={"number"}
          value={bodyMass}
          onChange={(event) => setBodyMass(event.target.value)}
        />
        <DatePicker
          id={"dateInput"}
          label={"Datum"}
          value={date}
          onChange={(value) => setDate(value ?? date)}
        />
        <DurationInput value={duration} onChange={setDuration} />
        <TextField
          id={"commentsInput"}
          label={"Kommentare"}
          multiline
          value={comments}
          onChange={(event) => setComments(event.target.value)}
        />
        <MedicationsInput value={medications} onChange={setMedications} />
      </DialogContent>
      <DialogActions>
        <Button color="primary" onClick={onAbort}>
          Abbrechen
        </Button>
        <Button color="primary" onClick={handleSubmit}>
          Speichern
        </Button>
      </DialogActions>
    </Dialog>
  );
}
