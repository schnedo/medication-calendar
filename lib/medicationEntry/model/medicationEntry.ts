import Medication from "./medication";
import { BodyMass } from "../../contact/model";
import Duration from "./duration";
import { v4 } from "uuid";

export default class MedicationEntry {
  readonly id: string = v4();

  constructor(
    readonly date: Date,
    readonly medications: Medication[],
    readonly duration: Duration,
    readonly comments: string,
    readonly bodyMass: BodyMass,
  ) {}
}
