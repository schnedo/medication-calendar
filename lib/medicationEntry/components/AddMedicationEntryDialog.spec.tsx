import { render, waitFor } from "@testing-library/react";
import AddMedicationEntryDialog from "./AddMedicationEntryDialog";
import { mock } from "@userlike/joke";
import userEvent from "@testing-library/user-event";
import { Duration, Medication, MedicationEntry } from "../model";
import { MuiPickersUtilsProvider } from "@material-ui/pickers";
import DateFnsUtils from "@date-io/date-fns";

const { default: DurationInput } = mock(import("./DurationInput"));
const { default: MedicationsInput } = mock(import("./MedicationsInput"));

describe("AddMedicationEntryDialog", () => {
  beforeEach(() => {
    DurationInput.mockReturnValue(<div>mocked Duration Input</div>);
    MedicationsInput.mockReturnValue(<div>mocked MedicationsInput</div>);
  });

  it("should render correctly", async () => {
    expect.hasAssertions();

    jest.setSystemTime(new Date(2021, 1, 13));

    const { container } = render(
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <AddMedicationEntryDialog open={true} />
      </MuiPickersUtilsProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  it("should call onSubmit when submitting", async () => {
    expect.hasAssertions();

    const onSubmit = jest.fn();

    jest.setSystemTime(new Date(2021, 1, 1));

    const duration: Duration = { minutes: 15, hours: 2 };
    DurationInput.mockImplementation(({ onChange }) => (
      <div data-testid={"durationInput"} onClick={() => onChange(duration)}>
        mocked DurationInput
      </div>
    ));

    const medications: Medication[] = [
      {
        id: "1",
        dose: {
          amount: 200,
        },
        medicament: {
          name: "name",
          batchNumber: {
            number: 12345,
          },
        },
      },
    ];
    MedicationsInput.mockImplementation(({ onChange }) => (
      <div
        data-testid={"medicationsInput"}
        onClick={() => onChange && onChange(medications)}
      >
        mocked MedicationsInput
      </div>
    ));

    const { getByRole, getByLabelText, getByTestId, getByText } = render(
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <AddMedicationEntryDialog open={true} onSubmit={onSubmit} />
      </MuiPickersUtilsProvider>,
    );

    await userEvent.type(
      getByLabelText("KÃ¶rpergewicht"),
      "{backspace}{backspace}10",
    );
    userEvent.click(getByLabelText("Datum"));
    userEvent.click(getByRole("button", { name: "10" }));
    userEvent.click(getByRole("button", { name: "OK" }));
    userEvent.click(getByTestId("durationInput"));
    await userEvent.type(getByLabelText("Kommentare"), "comments");
    userEvent.click(getByTestId("medicationsInput"));
    userEvent.click(getByText("Speichern"));

    // await waitFor(() => {
    //   expect(onSubmit).toHaveBeenCalledTimes(1);
    // });
    // const expected: MedicationEntry = {
    //   bodyMass: {
    //     amount: 10,
    //   },
    //   comments: "comments",
    //   date: new Date(2021, 1, 10),
    //   duration: { minutes: 15, hours: 2 },
    //   id: "",
    //   medications,
    // };
    // expect(onSubmit).toHaveBeenLastCalledWith(expected);
  });

  it("should call onAbort when aborting", async () => {
    expect.hasAssertions();

    const onAbort = jest.fn();

    jest.setSystemTime(new Date(2021, 1, 1));

    const { getByRole } = render(
      <MuiPickersUtilsProvider utils={DateFnsUtils}>
        <AddMedicationEntryDialog open={true} onAbort={onAbort} />
      </MuiPickersUtilsProvider>,
    );

    userEvent.click(getByRole("button", { name: "Abbrechen" }));
    await waitFor(() => {
      expect(onAbort).toHaveBeenCalledTimes(1);
    });
  });
});
