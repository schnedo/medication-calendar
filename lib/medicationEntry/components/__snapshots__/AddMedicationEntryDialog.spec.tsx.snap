// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMedicationEntryDialog should render correctly 1`] = `
<div
  aria-hidden="true"
/>
`;
